File: zero-gateway/pkg/circuitbreaker/circuit_breaker.go
package circuitbreaker

import (
	"sync"
	"time"
)

// State 熔断器状态
type State string

const (
	StateClosed   State = "closed"    // 关闭：正常状态，允许请求通过
	StateOpen     State = "open"      // 打开：熔断状态，拒绝请求
	StateHalfOpen State = "half_open" // 半开：尝试恢复，允许少量请求通过
)

// CircuitBreaker 熔断器实现
type CircuitBreaker struct {
	name             string
	failureThreshold int           // 连续失败次数阈值
	successThreshold int           // 半开状态下成功次数阈值（用于恢复）
	timeout          time.Duration // 熔断持续时间

	state           State
	failureCount    int
	successCount    int
	lastFailureTime *time.Time
	mu              sync.RWMutex
}

// Config 熔断器配置
type Config struct {
	FailureThreshold int           // 触发熔断的连续失败次数
	SuccessThreshold int           // 半开状态下需要连续成功的次数
	Timeout          time.Duration // 熔断持续时间
}

// DefaultConfig 返回默认配置
func DefaultConfig() Config {
	return Config{
		FailureThreshold: 5,
		SuccessThreshold: 2,
		Timeout:          60 * time.Second,
	}
}

// NewCircuitBreaker 创建新的熔断器
func NewCircuitBreaker(name string, config Config) *CircuitBreaker {
	return &CircuitBreaker{
		name:             name,
		failureThreshold: config.FailureThreshold,
		successThreshold: config.SuccessThreshold,
		timeout:          config.Timeout,
		state:            StateClosed,
		failureCount:     0,
		successCount:     0,
	}
}

// Allow 检查是否允许请求通过
func (cb *CircuitBreaker) Allow() bool {
	cb.mu.Lock()
	defer cb.mu.Unlock()

	now := time.Now()

	// 如果处于打开状态，检查是否超时
	if cb.state == StateOpen {
		if cb.lastFailureTime != nil && now.Sub(*cb.lastFailureTime) >= cb.timeout {
			// 超时，进入半开状态
			cb.state = StateHalfOpen
			cb.successCount = 0
			return true
		}
		// 仍在熔断期内，拒绝请求
		return false
	}

	// 关闭或半开状态，允许请求
	return true
}

// OnSuccess 记录成功
func (cb *CircuitBreaker) OnSuccess() {
	cb.mu.Lock()
	defer cb.mu.Unlock()

	if cb.state == StateHalfOpen {
		cb.successCount++
		if cb.successCount >= cb.successThreshold {
			// 连续成功达到阈值，恢复关闭状态
			cb.state = StateClosed
			cb.failureCount = 0
			cb.successCount = 0
		}
	} else {
		// 关闭状态下，重置失败计数
		cb.failureCount = 0
	}
}

// OnFailure 记录失败
func (cb *CircuitBreaker) OnFailure() {
	cb.mu.Lock()
	defer cb.mu.Unlock()

	now := time.Now()
	cb.lastFailureTime = &now

	if cb.state == StateHalfOpen {
		// 半开状态下失败，立即重新打开
		cb.state = StateOpen
		cb.successCount = 0
	} else {
		// 关闭状态下，增加失败计数
		cb.failureCount++
		if cb.failureCount >= cb.failureThreshold {
			// 达到失败阈值，打开熔断器
			cb.state = StateOpen
		}
	}
}

// GetState 获取当前状态
func (cb *CircuitBreaker) GetState() State {
	cb.mu.RLock()
	defer cb.mu.RUnlock()
	return cb.state
}

// Reset 手动重置熔断器
func (cb *CircuitBreaker) Reset() {
	cb.mu.Lock()
	defer cb.mu.Unlock()
	cb.state = StateClosed
	cb.failureCount = 0
	cb.successCount = 0
	cb.lastFailureTime = nil
}

// GetStats 获取统计信息
func (cb *CircuitBreaker) GetStats() map[string]interface{} {
	cb.mu.RLock()
	defer cb.mu.RUnlock()

	stats := map[string]interface{}{
		"name":              cb.name,
		"state":             string(cb.state),
		"failure_count":     cb.failureCount,
		"success_count":     cb.successCount,
		"failure_threshold": cb.failureThreshold,
		"success_threshold": cb.successThreshold,
		"timeout":           cb.timeout.String(),
	}

	if cb.lastFailureTime != nil {
		stats["last_failure_time"] = cb.lastFailureTime.Format(time.RFC3339)
	}

	return stats
}

File: zero-agent/core/circuit_breaker.py
"""
熔断器 - 用于保护上游服务，避免持续失败导致资源浪费
"""

import time
import logging
from enum import Enum
from typing import Optional
from threading import Lock

logger = logging.getLogger(__name__)


class CircuitState(Enum):
    """熔断器状态"""
    CLOSED = "closed"  # 关闭：正常状态，允许请求通过
    OPEN = "open"  # 打开：熔断状态，拒绝请求
    HALF_OPEN = "half_open"  # 半开：尝试恢复，允许少量请求通过


class CircuitBreaker:
    """熔断器实现
    
    使用滑动窗口统计失败率，当失败率超过阈值时触发熔断。
    """
    
    def __init__(
        self,
        failure_threshold: int = 5,  # 连续失败次数阈值
        success_threshold: int = 2,  # 半开状态下成功次数阈值（用于恢复）
        timeout: float = 60.0,  # 熔断持续时间（秒）
        name: str = "default"
    ):
        """
        初始化熔断器
        
        Args:
            failure_threshold: 触发熔断的连续失败次数
            success_threshold: 半开状态下需要连续成功的次数才能恢复
            timeout: 熔断持续时间，超过此时间后进入半开状态
            name: 熔断器名称，用于日志
        """
        self.failure_threshold = failure_threshold
        self.success_threshold = success_threshold
        self.timeout = timeout
        self.name = name
        
        self._state = CircuitState.CLOSED
        self._failure_count = 0
        self._success_count = 0
        self._last_failure_time: Optional[float] = None
        self._lock = Lock()
        
        logger.info(
            f"熔断器 [{name}] 初始化: "
            f"失败阈值={failure_threshold}, "
            f"成功阈值={success_threshold}, "
            f"超时={timeout}秒"
        )
    
    def call(self, func, *args, **kwargs):
        """同步调用，带熔断保护"""
        if not self._allow_request():
            raise CircuitBreakerOpenError(
                f"熔断器 [{self.name}] 处于打开状态，请求被拒绝"
            )
        
        try:
            result = func(*args, **kwargs)
            self._on_success()
            return result
        except Exception as e:
            self._on_failure()
            raise
    
    async def call_async(self, func, *args, **kwargs):
        """异步调用，带熔断保护"""
        if not self._allow_request():
            raise CircuitBreakerOpenError(
                f"熔断器 [{self.name}] 处于打开状态，请求被拒绝"
            )
        
        try:
            result = await func(*args, **kwargs)
            self._on_success()
            return result
        except Exception as e:
            self._on_failure()
            raise
    
    def _allow_request(self) -> bool:
        """检查是否允许请求通过"""
        with self._lock:
            now = time.time()
            
            # 如果处于打开状态，检查是否超时
            if self._state == CircuitState.OPEN:
                if self._last_failure_time and (now - self._last_failure_time) >= self.timeout:
                    # 超时，进入半开状态
                    self._state = CircuitState.HALF_OPEN
                    self._success_count = 0
                    logger.info(f"熔断器 [{self.name}] 超时，进入半开状态，允许测试请求")
                    return True
                else:
                    # 仍在熔断期内，拒绝请求
                    return False
            
            # 关闭或半开状态，允许请求
            return True
    
    def _on_success(self):
        """记录成功"""
        with self._lock:
            if self._state == CircuitState.HALF_OPEN:
                self._success_count += 1
                if self._success_count >= self.success_threshold:
                    # 连续成功达到阈值，恢复关闭状态
                    self._state = CircuitState.CLOSED
                    self._failure_count = 0
                    self._success_count = 0
                    logger.info(f"熔断器 [{self.name}] 恢复，进入关闭状态")
            else:
                # 关闭状态下，重置失败计数
                self._failure_count = 0
    
    def _on_failure(self):
        """记录失败"""
        with self._lock:
            self._last_failure_time = time.time()
            
            if self._state == CircuitState.HALF_OPEN:
                # 半开状态下失败，立即重新打开
                self._state = CircuitState.OPEN
                self._success_count = 0
                logger.warning(f"熔断器 [{self.name}] 半开状态下失败，重新打开")
            else:
                # 关闭状态下，增加失败计数
                self._failure_count += 1
                if self._failure_count >= self.failure_threshold:
                    # 达到失败阈值，打开熔断器
                    self._state = CircuitState.OPEN
                    logger.warning(
                        f"熔断器 [{self.name}] 触发熔断，"
                        f"连续失败 {self._failure_count} 次"
                    )
    
    def get_state(self) -> CircuitState:
        """获取当前状态"""
        with self._lock:
            return self._state
    
    def reset(self):
        """手动重置熔断器"""
        with self._lock:
            self._state = CircuitState.CLOSED
            self._failure_count = 0
            self._success_count = 0
            self._last_failure_time = None
            logger.info(f"熔断器 [{self.name}] 已手动重置")
    
    def get_stats(self) -> dict:
        """获取统计信息"""
        with self._lock:
            return {
                "name": self.name,
                "state": self._state.value,
                "failure_count": self._failure_count,
                "success_count": self._success_count,
                "last_failure_time": self._last_failure_time,
                "failure_threshold": self.failure_threshold,
                "success_threshold": self.success_threshold,
                "timeout": self.timeout
            }


class CircuitBreakerOpenError(Exception):
    """熔断器打开时抛出的异常"""
    pass
